src/Common/src/System/Buffers/ArrayBufferWriter.cs

```diff
@@ -0,0 +1,187 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+
+namespace System.Buffers
+{
+    /// <summary>
+    /// Represents a heap-based, array-backed output sink into which <typeparam name="T"/> data can be written.
+    /// </summary>
+#if USE_ABW_INTERNALLY
+    internal
+#else
+    public
+#endif
+    sealed class ArrayBufferWriter<T> : IBufferWriter<T>
```

Comment:
Add `where T:Struct` to avoid mixing objects and memory.

Reply:
I would like to match the `IBufferWriter<T>` constraint here (i.e. none). If there is strong opposition, we should discuss both outside of this PR in an API review.

Reply:
I'd rather not artificially constrain it, unless there's good reason to.

Reply:
Calling `GetMemory()` would be a bit odd with an array of objects. I'm assuming we don't do anything unsafe with these object pointers like copying (in the middle of a gc).

Reply:
> Calling GetMemory() would be a bit odd with an array of objects.

Why?

Reply:
The scenario is a bit edge case IMO - when would you pass a writeable buffer of objects? There's an indirection in memory (element->object) and "ownership" now becomes a bit more ambiguous - i.e. what if the objects implement IDispose -- who's going to call that?

Reply:
It's no different than storing objects into an array; when do you pass arrays of things, like strings? You also can have the exact same ownership question with structs, which might themselves have resources that have lifetime, or wrap any number of reference types. Limiting it is just an artificial restriction.

Reply:
I agree that limiting isn't a good thing if we don't need to.

Since Memory<T> has lifetime semantics (don't hold onto it after your method returns etc) it is not the same as an array. 

Consider a buffer type that is only for data (or primitives like char or byte - not objects) and in that case there could be some memory clear and copy-like efficiencies regarding clearing after use and shuffling unprocessed elements to the beginning when making room for more data. Originally I was thinking we may be doing some of these operations here which would make it unsafe for objects...